BUGFIX: Support ORM attributes

kdambekalns opened the pull request:
Hopefully fixes #2602 and allows to use PHP attributes to configure Doctrine ORM, not only annotations like before.

**Upgrade instructions**

No change should be needed.

**Checklist**

- [x] Code follows the PSR-2 coding style
- [x] Tests have been created, run and adjusted as needed
- [x] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [x] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [x] Reviewer - The first section explains the change briefly for change-logs
- [ ] ~~Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions~~

kdambekalns committed:
TASK: Clean up persistence tests code

# Conflicts:
#	Neos.Flow/Tests/Functional/Persistence/Aspect/PersistenceMagicAspectTest.php

kdambekalns committed:
TASK: Tiny code cleanup

mficzel committed:
TASK: Duplicate Test and Fixtures ... no php8 adjustments yet

This is here to demonstrate that the tests can succeed and are a 1:1 copy of the classic fixtures.

kdambekalns commented:
**This does not yet work** – whether this "minimally invasive" approach is feasible is yet to be proven…

kdambekalns committed:
BUGFIX: Use custom reader to access ORM configuration

The Doctrine reader exclusively used annotations, our reflection service
also takes attributes into account.

This should allow to use both for configuring ORM.

kdambekalns committed:
BUGFIX: Handle proxy classnames correctly

kdambekalns committed:
TASK: Update minimum doctrine/common 3.0.3 to 3.1.0

kdambekalns commented:
@mficzel Using the reflection service in our own reader implementation works. 

If our assumptions are correct, new tests should pass, too.

mficzel committed:
TASK: Adjust PHP8 persistence test-fixtures to php 8 syntax

- ORM annotations > ORM attributes
- Flow annotation > Flow attributes
- @param and @return annotations to parameter and return types

!!! Nested attributes are not supported in php 8.0 so they stay annotations for now !!!

mficzel commented:
@kdambekalns something is still fishy ...
```
Doctrine\ORM\Mapping\MappingException: No identifier/primary key specified for Entity "Neos\Flow\Tests\Functional\Persistence\FixturesPHP8\AnnotatedIdEntity" sub class of "Neos\Flow\Tests\Functional\Persistence\FixturesPHP8\AnnotatedIdEntity_Original". Every Entity must have an identifier/primary key.

/var/www/html/Packages/Libraries/doctrine/orm/src/Mapping/MappingException.php:44
/var/www/html/Packages/Libraries/doctrine/orm/src/Mapping/ClassMetadataInfo.php:1194
/var/www/html/Packages/Libraries/doctrine/orm/src/Mapping/ClassMetadataFactory.php:284
/var/www/html/Packages/Libraries/doctrine/orm/src/Mapping/ClassMetadataFactory.php:264
/var/www/html/Packages/Libraries/doctrine/persistence/src/Persistence/Mapping/AbstractClassMetadataFactory.php:343
/var/www/html/Packages/Libraries/doctrine/persistence/src/Persistence/Mapping/AbstractClassMetadataFactory.php:207
/var/www/html/Packages/Libraries/doctrine/persistence/src/Persistence/Mapping/AbstractClassMetadataFactory.php:96
/var/www/html/Data/Temporary/Testing/Cache/Code/Flow_Object_Classes/Neos_Flow_Persistence_Doctrine_PersistenceManager.php:317
/var/www/html/Packages/Framework/Neos.Flow/Tests/FunctionalTestCase.php:159

```

mficzel commented:
I wonder wether we also have to adjust `\Neos\Flow\Persistence\Doctrine\Mapping\ClassMetadata->_initializeReflection`

kdambekalns commented:
Making `@Flow\Entity` an attribute in "real code" within the Flow classes allows a clean compile run. And `#[ORM\Id]` and `#[Flow\Identity]` do work, too. So, is only testing affected?

kdambekalns commented:
Doh… I removed all the adjusted entities except `AnnotatedIdEntity` to limit the scope. Turns out **this is broken**:
```php
#[Flow\Entity]
class AnnotatedIdEntity
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\SequenceGenerator(sequenceName: 'annotatedidentity_seq')]
    protected string $id;
```

But **this does work**:
```php
#[Flow\Entity]
class AnnotatedIdEntity
{
    /** @var string */
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\SequenceGenerator(sequenceName: 'annotatedidentity_seq')]
    protected string $id;
```

The missing `@var` breaks it! On to digging deeper…

mficzel commented:
We should think about targeting Neos 8.4 + PHP 8.2 as i noticed that in Neos 8.3 with PHP 8.0 we: 

- Cannot declare nested attributes in PHP as those are a PHP 8.1 feature
- Even with PHP 8.2 + Neos 8.3 Nested Attributes are destroyed by the proxy class builder. I hope this is better in 8.4

mficzel commented:
Checked Neos 8.4 and the proxy builder still destroys nested attributes. Created #3511 for that.

kdambekalns committed:
TASK: Code cleanup

kdambekalns commented:
I am confused. Locally this PR passes on PHP 8.3 if I add this:
[var-annotations.patch](https://github.com/user-attachments/files/23096789/var-annotations.patch)

How does this relate to broken attribute rendering?

- In the sense it will never work on PHP 8.0 and 8.1, and the solution is to go for Flow 8.4, as that requires PHP 8.2? Fine with me!
- But that will not automagically fix the "missing type issue", or is that something that has changed in Flow 8.4 already?

mficzel commented:
I tried rebasing this pr to 8.4 and together with the changes from #3512 i could successfully define a repository and entity via attributes. 

However the properties in the entity were not detected unless i added the @var annotation. So this problem persists even in 8.4. 

However it looks like it might really be the last real problem as the proxy classes were successfully created otherwise and flow did not complain about anything.

kdambekalns commented:
> However it looks like it might really be the last real problem as the proxy classes were successfully created otherwise and flow did not complain about anything.

That's great news already!

kdambekalns committed:
BUGFIX: Respect type declaration when building class-schema

Without this change, any class property is ignored, if it has no `@var`
tag. As soon as a property lacks that tag, the early return in method
`evaluateClassPropertyAnnotationsForSchema` kicks in.

When using type declarations, this is nonsense, though. This change
thus makes sure the type is fetched from the native type declaration
if no `@var` tag is found. If a `@var` tag exists, it takes precedence.
This allows to do things like this to declare typed collections:

```php
/**
 * @var Collection<AnnotatedIdentitiesEntity>
 */
#[ORM\ManyToMany(indexBy: 'author')]
protected Collection $annotatedIdentitiesEntities;
```

kdambekalns committed:
TASK: Adjust (new) fixture to Flow 8.4

The ``json_array`` type is removed in Doctrine DBAL 3.0.

kdambekalns committed:
TASK: Make compatible with lowest dependency

kdambekalns committed:
TASK: Raise minimal doctrine/orm version from 2.12 to 2.17

This allows complex return types like `ArrayCollection|Collection` or
`static`.

kdambekalns commented:
🎉

mhsdesign commented on Neos.Flow/Tests/Functional/Persistence/FixturesPHP8/Post.php:
```diff
@@ -0,0 +1,126 @@
+    /**
+     * Still annotated because nested Attributes require PHP 8.1
+     * @todo Adjust to attributes for Flow 8.4 when the min PHP version will be 8.1
+     *
+     * @var Collection<Post>
+     * @ORM\ManyToMany
+     * @ORM\JoinTable(inverseJoinColumns={@ORM\JoinColumn(name="related_post_id")})
+     */
+    protected Collection $related;
```
shouldn't we now be able to use nested attributes here? From https://github.com/neos/flow-development-collection/pull/3512

kitsunet replied:
I think this prevents it: https://github.com/neos/flow-development-collection/issues/3511

kdambekalns replied:
But #5312 fixes #5311…

mficzel replied:
@mhsdesign yes if #3512 is in we should be able to adjust this. Before this caused broken php code in the proxy classes.

kdambekalns replied:
Yay, tests pass even with nested attributes. 🎉

mficzel replied:
Adjusted the tests ... still green

mficzel committed:
TASK: Convert nested orm annotations to attributes

kitsunet commented on Neos.Flow/Classes/Reflection/ReflectionService.php:
```diff
@@ -1568,16 +1568,19 @@ protected function evaluateClassPropertyAnnotationsForSchema(ClassSchema $classS
         if (count($varTagValues) > 1) {
             throw new InvalidPropertyTypeException('More than one @var annotation given for "' . $className . '::$' . $propertyName . '"', 1367334366);
         }
+        if (count($varTagValues) === 0) {
```
Do we want to prefer the var tags over the declaredType if both exist? 
Spontanously would have expected something like this:

$declaredType = $this->getPropertyType($className, $propertyName);
if ($declaredType = null && count($varTagValues) === 1)

mficzel replied:
We need var tags for typed arrays or collections as those cannot be expressed natively so i assume they will have to take precedence.

kdambekalns replied:
Ex-act-ly what @mficzel said!

kdambekalns replied:
Added a comment to clarify that for us and future generations…

kitsunet commented:
I left one question but it's not a blocker for merge I guess...

kdambekalns committed:
TASK: Explain var tag precedence in ReflectionService

Added comment to clarify the need for `@var` tags on typed arrays or collections.
